samples/apps/local_commit_handler/localcommithandler.cpp

```diff
@@ -0,0 +1,149 @@
+        ctx.rpc_ctx->set_response_header("mytxid", txid.to_str());
+
+        Response& resp =
+          *static_cast<Response*>(ctx.rpc_ctx->get_response_user_data());
+
+        throw std::runtime_error("oops, might have failed serialization");
+
+        resp.tx_id = txid.to_str();
+
+        ctx.rpc_ctx->set_response_body(nlohmann::json(resp).dump());
+      };
+
+      auto increment = [this](auto& ctx) {
+        auto counter_handle = ctx.tx.template rw<CounterMap>("counters");
+
+        auto current_value = counter_handle->get("counter").value_or(0);
+
+        auto new_value = current_value + 1;
+        counter_handle->put("counter", new_value);
+
+        Response resp;
+        resp.value = new_value;
+
+        ctx.rpc_ctx->set_response_status(HTTP_STATUS_OK);
+        ctx.rpc_ctx->set_response_header(
+          http::headers::CONTENT_TYPE, http::headervalues::contenttype::JSON);
+        ctx.rpc_ctx->set_response_user_data(static_cast<Response*>(&resp));
+      };
+      make_endpoint_with_commit_handler(
+        "/increment", HTTP_POST, increment, add_tx_id, ccf::no_auth_required)
+        .set_auto_schema<void, Response>()
+        .install();
+
+      make_endpoint_with_commit_handler(
+        "/increment_exception",
+        HTTP_POST,
+        increment,
+        add_tx_id_exception,
+        ccf::no_auth_required)
+        .set_auto_schema<void, Response>()
+        .install();
+
+      auto decrement = [this](auto& ctx) {
+        auto counter_handle = ctx.tx.template rw<CounterMap>("counters");
+
+        auto current_value = counter_handle->get("counter").value_or(0);
+
+        auto new_value = current_value - 1;
+        counter_handle->put("counter", new_value);
+
+        Response resp;
+        resp.value = new_value;
+
+        ctx.rpc_ctx->set_response_status(HTTP_STATUS_OK);
+        ctx.rpc_ctx->set_response_header(
+          http::headers::CONTENT_TYPE, http::headervalues::contenttype::JSON);
+        ctx.rpc_ctx->set_response_user_data(static_cast<Response*>(&resp));
+      };
+      make_endpoint_with_commit_handler(
+        "/decrement", HTTP_POST, decrement, add_tx_id, ccf::no_auth_required)
+        .set_auto_schema<void, Response>()
+        .install();
+
+      make_endpoint_with_commit_handler(
+        "/decrement_exception",
+        HTTP_POST,
+        decrement,
+        add_tx_id_exception,
+        ccf::no_auth_required)
+        .set_auto_schema<void, Response>()
+        .install();
+
+      auto value = [this](auto& ctx) {
+        auto counter_handle = ctx.tx.template ro<CounterMap>("counters");
+
+        auto current_value = counter_handle->get("counter").value_or(0);
+
+        Response resp;
+        resp.value = current_value;
+
+        ctx.rpc_ctx->set_response_status(HTTP_STATUS_OK);
+        ctx.rpc_ctx->set_response_header(
+          http::headers::CONTENT_TYPE, http::headervalues::contenttype::JSON);
+        ctx.rpc_ctx->set_response_user_data(static_cast<Response*>(&resp));
```

Comment:
This won't transfer ownership of `resp`, which is just a local variable which will be deleted  at line 123. The `void*` stored in `rpc_ctx` is then pointing at junk - part of the stack that previously held `resp`.

This should instead be an owning pointer type so that we can extend the lifetime to the future call, and transfer ownership with this call. Either a `unique_ptr` or `shared_ptr`.

Reply:
Is there any best practice for `unique_ptr` vs `shared_ptr` here given that they can still just get the value back straight away?

Currently I've gone with a `unique_ptr` so that the rpc context actually owns it.

Reply:
No strong reason either way. I think while we expect single-ownership, it's simpler to make it a `unique_ptr`. Then iff we find a reason for it to be a `shared_ptr` before ownership is transferred, we can make it a `shared_ptr`.